### PR TITLE
Internal Json: Adds String Encoding For Binary

### PR DESCRIPTION
**Pull request by bchong95**
# Internal Json: Adds String Encoding For Binary

## Description

Porting the backend improvements made to string which are (https://msdata.visualstudio.com/CosmosDB/_git/CosmosDB/pullrequest/421829): 

>* String References
>Similar to string dictionary encoding, we maintain an internal dictionary of written strings, that they meet certain criteria, and write repeated strings as a reference to the first written one. String references are written as a byte-marker and an offset, where the offset is either 1, 2, 3, or 4 bytes.

>* GUID Values
>Strings that we identify as GUIDs are encoded as 16-byte binary values rather than a 36-byte text value. We differentiate between lowercase and uppercase letters and do not allow encoding of mixed casing values.

>* Character-set Encoding
>For strings with a finite set of characters, for instance date-time and hexadecimal values, we encode each character as a position within this character set. For 16-byte character sets, each character is encoded as 4 bits which provides 50% reduction in size.

>* String Compression
>We first identify the range of characters of a string, and depending on the length of this range we encode each character it as either 4, 5, 6 or 7 bits.

Some highlights in perf:

Storage numbers from the backend:

```
--------------------------------------------------------------------------------------
Encoding                             | Before (bytes)        | After (bytes)         |
--------------------------------------------------------------------------------------
Text                                 | 436,710,051                                   |
Cosmos Binary                        | 378,739,143     0.84  | 227,386,067     0.49  |
Cosmos Binary w/ Dictionary Encoding | 298,399,964     0.66  | 215,723,247     0.46  |
--------------------------------------------------------------------------------------
```
 
Binary to Binary RoundTrip: 
* 99.15 us to 31.93 us 
* 688.04 KB to 246.35 KB

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Json/JsonBinaryEncoding.Chars.cs [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+namespace Microsoft.Azure.Cosmos.Json
+{
+    using System;
+    using System.Collections;
+    using System.Collections.Immutable;
+    using Microsoft.Azure.Documents;
+
+    internal static partial class JsonBinaryEncoding
+    {
+        public readonly struct Chars
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Json/JsonBinaryEncoding.Chars.cs:14]: >Chars [](start = 31, length = 5)

reconsider the name here

Comment (sboshra) [reply Microsoft.Azure.Cosmos/src/Json/JsonBinaryEncoding.Chars.cs:14]: StringCompressionLookupTables

---
In reply to: [497181094](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/1880#discussion_r497181094) [](ancestors = 497181094)

File: Microsoft.Azure.Cosmos/src/Json/JsonBinaryEncoding.cs [hunk 2]
```diff
@@ -218,5 +240,50 @@ public static bool TryGetValueLength(ReadOnlySpan<byte> buffer, out int length)
...
+        {
+            Span<TFixedType> bufferAsFixedType = MemoryMarshal.Cast<byte, TFixedType>(buffer);
+            bufferAsFixedType[0] = value;
+        }
+
+        public static TFixedType GetFixedSizedValue<TFixedType>(ReadOnlySpan<byte> buffer)
+            where TFixedType : struct
+        {
+            return MemoryMarshal.Cast<byte, TFixedType>(buffer)[0];
+        }
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Json/JsonBinaryEncoding.cs:255]: try avoid Marshal and opt for different overloads for different types

File: Microsoft.Azure.Cosmos/src/Json/JsonBinaryEncoding.Strings.cs [hunk 3]
```diff
@@ -5,46 +5,236 @@
 namespace Microsoft.Azure.Cosmos.Json
 {
     using System;
+    using System.Collections;
+    using System.Collections.Immutable;
+    using System.Linq;
     using System.Runtime.InteropServices;
     using Microsoft.Azure.Cosmos.Core.Utf8;
+    using Microsoft.Azure.Cosmos.Rntbd;
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Json/JsonBinaryEncoding.Strings.cs:13]: double check this one

File: Microsoft.Azure.Cosmos/src/Json/JsonBinaryEncoding.Strings.cs [hunk 4]
```diff
@@ -5,46 +5,236 @@
...
 
     internal static partial class JsonBinaryEncoding
     {
+        public const int GuidLength = 36;
+        public const int EncodedGuidLength = 17;
+        public const int GuidWithQuotesLength = GuidLength + 2;
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Json/JsonBinaryEncoding.Strings.cs:19]: >GuidWithQuotesLength [](start = 25, length = 20)

move it right below GuidLength

File: Microsoft.Azure.Cosmos/src/Json/JsonBinaryEncoding.Strings.cs [hunk 5]
```diff
@@ -5,46 +5,236 @@
...
+        private const int MinCompressedStringLength = Min4BitCharSetStringLength;
+
         /// <summary>
-        /// Gets the string value from the binary reader.
+        ///  Determines whether a type-marker is potentially for a buffered string value
         /// </summary>
-        /// <param name="stringToken">The buffer that has the string.</param>
-        /// <param name="jsonStringDictionary">The JSON string dictionary.</param>
-        /// <returns>A string value from the binary reader.</returns>
+        private static readonly ImmutableArray<bool> IsPotentiallyBufferedString = new bool[]
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Json/JsonBinaryEncoding.Strings.cs:32]: >IsPotentiallyBufferedString [](start = 53, length = 27)

IsBufferedStringCandidate

File: Microsoft.Azure.Cosmos/src/Json/JsonBinaryEncoding.Strings.cs [hunk 6]
```diff
@@ -241,6 +443,57 @@ private static bool TryGetUserStringId(Utf8Span stringToken, out int userStringI
...
+                            value = default;
+                            return false;
+                        }
+
+                        return TryGetBufferedStringValue(
+                            buffer,
+                            buffer.Slice(start: MemoryMarshal.Read<ushort>(stringTokenSpan)),
+                            jsonStringDictionary,
+                            out value);
+                    case JsonBinaryEncoding.TypeMarker.ReferenceString3ByteOffset:
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Json/JsonBinaryEncoding.Strings.cs:471]: >case JsonBinaryEncoding.TypeMarker.ReferenceString3ByteOffset: [](start = 20, length = 62)

new line

File: Microsoft.Azure.Cosmos/src/Json/JsonBinaryEncoding.Strings.cs [hunk 7]
```diff
@@ -357,5 +610,816 @@ private static bool TryGetUserStringId(Utf8Span stringToken, out int userStringI
...
+                    return false;
+                }
+
+                valueCharSet.Set(charValue, true);
+            }
+
+            int firstSetBit = 0;
+            for (; (firstSetBit < valueCharSet.Length) && !valueCharSet[firstSetBit]; firstSetBit++)
+            {
+            }
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Json/JsonBinaryEncoding.Strings.cs:737]: move it as an extension method

File: Microsoft.Azure.Cosmos/src/Json/JsonBinaryEncoding.Strings.cs [hunk 8]
```diff
@@ -357,5 +610,816 @@ private static bool TryGetUserStringId(Utf8Span stringToken, out int userStringI
...
+
+            int charRange = lastSetBit - firstSetBit + 1;
+            int charCount = 0;
+            for (int i = 0; i < valueCharSet.Length; i++)
+            {
+                if (valueCharSet[i])
+                {
+                    charCount++;
+                }
+            }
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Json/JsonBinaryEncoding.Strings.cs:752]: consider a single loop to get all three values

File: Microsoft.Azure.Cosmos/src/Json/JsonBinaryEncoding.Strings.cs [hunk 9]
```diff
@@ -357,5 +610,816 @@ private static bool TryGetUserStringId(Utf8Span stringToken, out int userStringI
...
+                    return JsonBinaryEncoding.ValueLengths.GetCompressedStringLength(stringToken[1], numberOfBits: 7);
+
+                case JsonBinaryEncoding.TypeMarker.Packed7BitStringLength2:
+                    return JsonBinaryEncoding.ValueLengths.GetCompressedStringLength(GetFixedSizedValue<ushort>(stringToken.Slice(1)), numberOfBits: 7);
+
+                case JsonBinaryEncoding.TypeMarker.LowercaseGuidString:
+                case JsonBinaryEncoding.TypeMarker.UppercaseGuidString:
+                case JsonBinaryEncoding.TypeMarker.DoubleQuotedLowercaseGuidString:
+                    return 16;
+                default:
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Json/JsonBinaryEncoding.Strings.cs:1200]: >efault: [](start = 17, length = 7)

new line

File: Microsoft.Azure.Cosmos/src/Json/JsonReader.JsonBinaryReader.cs [hunk 10]
```diff
@@ -36,12 +37,39 @@ abstract partial class JsonReader : IJsonReader
...
+            JsonTokenType.String,  // <empty> 0x70
+            JsonTokenType.String,  // <empty> 0x71
+            JsonTokenType.String,  // <empty> 0x72
+            JsonTokenType.String,  // <empty> 0x73
+            JsonTokenType.String,  // <empty> 0x74
+            JsonTokenType.String,  // StrGL (Lowercase GUID string)
+            JsonTokenType.String,  // StrGU (Uppercase GUID string)
+            JsonTokenType.String,  // StrGQ (Double-quoted lowercase GUID string)
+
+            // Compressed strings [JsonTokenType.Stringx78, JsonTokenType.Stringx8JsonTokenType.String)
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Json/JsonReader.JsonBinaryReader.cs:63]: >Stringx78 [](start = 49, length = 9)

typo

**Review by sboshra — APPROVED** [review]
:shipit:

File: Microsoft.Azure.Cosmos/src/Json/JsonBinaryEncoding.cs (hunk 2 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Json/JsonBinaryEncoding.cs:255]: I looked into the code. The span is backed by a pointer and this operation is efficient.

File: Microsoft.Azure.Cosmos/src/Json/JsonBinaryEncoding.Strings.cs (hunk 7 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Json/JsonBinaryEncoding.Strings.cs:737]: not needed anymore since we fused the loops.

**Review by sboshra — APPROVED** [review]

**Pull request merged**